src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs

```diff
@@ -0,0 +1,135 @@
+
+                // This multiplication can overflow, so cast to uint first.
+                int indexableLength = (int)((uint)_array.Length * 32 - 1);
+
+                Debug.Assert(indexableLength >= -1, $"maximum indexable length: {indexableLength}");
+
+                return indexableLength;
+            }
+        }
+
+        public bool this[int index]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                Debug.Assert(!IsDefault);
+                Debug.Assert(index >= 0, $"Get - Negative - index: {index}, arrayLength: {_array.Length}");
+
+                int elementIndex = Div32Rem(index, out int extraBits);
+
+                Debug.Assert(elementIndex < _array.Length, $"Get - index: {index}, elementIndex: {elementIndex}, arrayLength: {_array.Length}, extraBits: {extraBits}");
+
+                return (_array[elementIndex] & (1 << extraBits)) != 0;
+            }
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            set
+            {
+                Debug.Assert(!IsDefault);
+                Debug.Assert(index >= 0, $"Set - Negative - index: {index}, arrayLength: {_array.Length}");
+
+                // Maximum possible array length if bitLength was int.MaxValue (i.e. 67_108_864)
+                Debug.Assert(_array.Length <= int.MaxValue / 32 + 1, $"index: {index}, arrayLength: {_array.Length}");
+
+                int elementIndex = Div32Rem(index, out int extraBits);
+
+                // Grow the array when setting a bit if it isn't big enough
+                // This way the caller doesn't have to check.
+                if (elementIndex >= _array.Length)
+                {
+                    // This multiplication can overflow, so cast to uint first.
+                    Debug.Assert(index >= 0 && index > (int)((uint)_array.Length * 32 - 1), $"Only grow when necessary - index: {index}, arrayLength: {_array.Length}");
+                    DoubleArray(elementIndex);
+                }
+
+                Debug.Assert(elementIndex < _array.Length, $"Set - index: {index}, elementIndex: {elementIndex}, arrayLength: {_array.Length}, extraBits: {extraBits}");
+
+                int newValue = _array[elementIndex];
+                if (value)
+                {
+                    newValue |= 1 << extraBits;
+                }
+                else
+                {
+                    newValue &= ~(1 << extraBits);
+                }
+                _array[elementIndex] = newValue;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void DoubleArray(int minSize)
+        {
+            Debug.Assert(_array.Length < int.MaxValue / 2, $"Array too large - arrayLength: {_array.Length}");
+            Debug.Assert(minSize >= 0 && minSize >= _array.Length);
+
+            int nextDouble = NextClosestPowerOf2(minSize + 1);
```

Comment:
You can do just: `Math.Max(minSize + 1, _array.Length * 2)`. No need for `NextClosestPowerOf2`.